cnxarchive/database.py

```diff
@@ -112,6 +113,57 @@ def initdb(settings):
             for schema_part in get_schema():
                 cursor.execute(schema_part)
 
+            db_host = db_connection.dsn.split(" ")[3].split("=")[1]
+            # If you are connecting to a database that is not localhost,
+            # don't initalize with virtualenv
+            if db_host == 'localhost' or db_host == '127.0.0.1':
+                # If virtualenv is active, use that for postgres
+                if hasattr(sys, 'real_prefix'):
+                    activate_path = os.path.join(
+                        os.path.realpath(sys.prefix),
+                        'bin/activate_this.py')
+                    cursor.execute("SELECT current_database();")
+                    db_name = cursor.fetchone()[0]
+
+                    cursor.execute("""SELECT schema_name FROM
+                                      information_schema.schemata
+                                      WHERE schema_name = 'venv';""")
+                    schema_exists = cursor.fetchall()
+
+                    if not schema_exists:
+                        cursor.execute("CREATE SCHEMA venv")
+                        cursor.execute("ALTER DATABASE \"{}\" SET "
+                                       "session_preload_libraries ="
+                                       "'session_exec'".format(db_name))
+                        cursor.execute("ALTER DATABASE \"{}\" SET "
+                                       "session_exec.login_name = "
+                                       "'venv.activate_venv'".format(db_name))
+                        sql = """CREATE FUNCTION venv.activate_venv() RETURNS
+void LANGUAGE plpythonu AS $_$
+import sys
+import os
+import site
+old_os_path = os.environ['PATH']
+os.environ['PATH'] = os.path.dirname(os.path.abspath('{activate_path}')) \
++ os.pathsep + old_os_path
+base = os.path.dirname(os.path.dirname(os.path.abspath('{activate_path}')))
+site_packages = os.path.join(base, 'lib', 'python%s' % sys.version[:3], \
+'site-packages')
+prev_sys_path = list(sys.path)
+import site
+site.addsitedir(site_packages)
+sys.real_prefix = sys.prefix
+sys.prefix = base
+# Move the added items to the front of the path:
+new_sys_path = []
+for item in list(sys.path):
+    if item not in prev_sys_path:
+        new_sys_path.append(item)
+        sys.path.remove(item)
+sys.path[:0] = new_sys_path
+$_$""".format(activate_path=activate_path)
```

Comment:
I don't know that I see anything wrong with porting this code here. I was originally suggesting you compile the activate_this module.

I believe that would look something like the following...

Before the sql function definition compile and marshal the module to bytecode.
```python
file = 'activate_this.py'
filepath = os.path.join('bin', file)
c = compile(open(file).read(), file, 'exec')
import marshal
bytecode = io.BytesIO()
marshal.dump(c, bytecode)
```
Then the bytecode is used to format the sql function, something like...
```python
filepath = "{filepath-formated-from-initdb}"
import marshal
bytecode = io.BytesIO("{bytecode-formated-from-initdb}")
c = marshal.load(bytecode)
exec(c, dict(__file__=filepath)
```

I don't know how well this will work without actually giving it a try. I think the ported version you came up with should work fine enough. Plus, the port will likely make it easier to read.